### ISSUE TITLE
How to do EMA update of the teacher model in accelerate?

### ISSUE THREAD

**Issue by HashmatShadab** [huggingface/accelerate#1056]
The template for the training of student and teacher model in  [https://github.com/microsoft/esvit/blob/main/main_esvit.py](url) has the following way to update teacher parameters during an epoch.
> ### def train_one_epoch(student, teacher, teacher_without_ddp, dino_loss, data_loader, optimizer, lr_schedule, wd_schedule, momentum_schedule, epoch, mixup_fn, p16_scaler, args):

> `# EMA update for the teacher

> with torch.no_grad():

>> m = momentum_schedule[it]  # momentum parameter
>> for param_q, param_k in zip(student.module.parameters(), teacher_without_ddp.parameters()):
> >>param_k.data.mul_(m).add_((1 - m) * param_q.detach().data)`

As done in [https://github.com/microsoft/esvit/blob/main/main_esvit.py](url) for distributed data parallel training. teacher_without_dpp is also passed during training for updating the teacher parameters manually, since no gradients are backpropogated through the teacher.

If I am using accelerate for distributing training:
student, teacher,  optimizer, data_loader= accelerator.prepare(student, teacher, optimizer, data_loader)
How should the teacher update look like?
> `# EMA update for the teacher

> with torch.no_grad():

>> m = momentum_schedule[it]  # momentum parameter
>> for param_q, param_k in zip(student.parameters(), teacher.parameters()):
> >>param_k.data.mul_(m).add_((1 - m) * param_q.detach().data)`

**Comment by sgugger** [comment]
cc @muellerzr

**Comment by muellerzr** [comment]
As `accelerate` lightly wraps around pytorch, it'd look such that:
```python
    student, teacher, optimizer, data_loader= accelerator.prepare(
        student, teacher, optimizer, data_loader
    )
    # EMA update for the teacher
    with torch.no_grad():
        m = momentum_schedule[it] # momentum parameter
        for param_q, param_k in zip(student.module.parameters(), teacher.module.parameters()):
            param_k.data.mul_(m).add_((1 - m) * param_q.detach().data)
```

However to fully recreate it the exact same (such that the teacher is not on DDP), your code must look like so:
```python
    student, optimizer, data_loader= accelerator.prepare(
        student, teacher, optimizer, data_loader
    )
    teacher.to(accelerator.device) # If they change the device of the teacher in the code
    # EMA update for the teacher
    with torch.no_grad():
        m = momentum_schedule[it] # momentum parameter
        for param_q, param_k in zip(student.module.parameters(), teacher.parameters()):
            param_k.data.mul_(m).add_((1 - m) * param_q.detach().data)
```
